src/dbnode/storage/index/aggregate_results.go

```diff
@@ -0,0 +1,206 @@
+	aggregateQueryOpts AggregateResultsOptions
+
+	resultsMap *AggregateResultsMap
+
+	idPool    ident.Pool
+	bytesPool pool.CheckedBytesPool
+
+	pool       AggregateResultsPool
+	valuesPool AggregateValuesPool
+}
+
+// NewAggregateResults returns a new AggregateResults object.
+func NewAggregateResults(
+	namespaceID ident.ID,
+	queryOpts QueryResultsOptions,
+	aggregateQueryOpts AggregateResultsOptions,
+	opts Options,
+) AggregateResults {
+	return &aggregatedResults{
+		nsID:               namespaceID,
+		queryOpts:          queryOpts,
+		aggregateQueryOpts: aggregateQueryOpts,
+		resultsMap:         newAggregateResultsMap(opts.IdentifierPool()),
+		idPool:             opts.IdentifierPool(),
+		bytesPool:          opts.CheckedBytesPool(),
+		pool:               opts.AggregateResultsPool(),
+		valuesPool:         opts.AggregateValuesPool(),
+	}
+}
+
+func (r *aggregatedResults) Reset(
+	nsID ident.ID,
+	queryOpts QueryResultsOptions,
+	aggregateQueryOpts AggregateResultsOptions,
+) {
+	r.Lock()
+
+	r.queryOpts = queryOpts
+	r.aggregateQueryOpts = aggregateQueryOpts
+
+	// finalize existing held nsID
+	if r.nsID != nil {
+		r.nsID.Finalize()
+	}
+
+	// make an independent copy of the new nsID
+	if nsID != nil {
+		nsID = r.idPool.Clone(nsID)
+	}
+	r.nsID = nsID
+
+	// reset all values from map first
+	for _, entry := range r.resultsMap.Iter() {
+		valueMap := entry.Value()
+		valueMap.reset()
+	}
+
+	// reset all keys in the map next
+	r.resultsMap.Reset()
+
+	// NB: could do keys+value in one step but I'm trying to avoid
+	// using an internal method of a code-gen'd type.
+	r.Unlock()
+}
+
+func (r *aggregatedResults) AddDocuments(batch []doc.Document) (int, error) {
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size := r.resultsMap.Len()
+	r.Unlock()
+	return size, err
+}
+
+func (r *aggregatedResults) addDocumentsBatchWithLock(
+	batch []doc.Document,
+) error {
+	for i := range batch {
+		err := r.addDocumentWithLock(batch[i])
+		if err != nil {
+			return err
+		}
+
+		if r.queryOpts.SizeLimit > 0 && r.resultsMap.Len() >= r.queryOpts.SizeLimit {
+			// Early return if limit enforced and we hit our limit.
+			break
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addDocumentWithLock(
+	document doc.Document,
+) error {
+	for _, field := range document.Fields {
+		if err := r.addFieldWithLock(field.Name, field.Value); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addFieldWithLock(
+	term []byte,
+	value []byte,
+) error {
+	// NB: can cast the []byte -> ident.ID to avoid an alloc
+	// before we're sure we need it.
+	var termID ident.ID = ident.BytesID(term)
+
+	// if this term hasn't been seen, ensure it should be included in output.
+	if !r.aggregateQueryOpts.TermFilter.Contains(termID) {
+		return nil
+	}
+
+	valueBytes := r.bytesPool.Get(len(value))
+	valueBytes.IncRef()
+	valueBytes.AppendAll(value)
+	valueBytes.DecRef()
+	valueID := r.idPool.BinaryID(valueBytes)
```

Comment:
Sorry, I realize you can actually use a helper function to make this code a little simpler:
```go
valueID := r.idPool.Clone(ident.BytesID(value))
```

Reply:
Nice 👍